### PR TITLE
feat: Add HalfDonut and QuarterDonut widgets

### LINKED ISSUE #2889: Create new widget 'HalfDonut'

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes #2889
Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

### CODE REVIEW

**Comment by jbrundage** [comment]
image

**Comment by jbrundage** [comment]
image

**Comment by jbrundage** [comment]
image

File: packages/chart/src/Pie.ts [hunk 1]
```diff
@@ -230,6 +235,10 @@ export class Pie extends SVGWidget {
         SVGWidget.prototype.exit.apply(this, arguments);
     }
 
+    getPieGenerator() {
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: This confused me at first - can you:
* Change to "updateD3Pie(): void"
* move the following code into it https://github.com/hpcc-systems/Visualization/blob/4eb8e4698f57a3d0955e990fd44644a034eee189/packages/chart/src/Pie.ts#L31-L37 into it
* Then call it at the start of the update method

Then the other classes can override the method as needed.

File: packages/chart/src/QuarterDonut.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+                    ;
+                break;
+        }
+    }
+    calcOuterRadius() {
+        const maxTextWidth = this.textSize(this.data().map(d => this.getLabelText({ data: d })), "Verdana", 12).width;
+        return Math.min(this._size.width - maxTextWidth - 10, this._size.height - 12 * 3) - 2;
+    }
+}
+QuarterDonut.prototype._class += " common_QuarterDonut";
```
Comment (GordonSmith) [inline packages/chart/src/QuarterDonut.ts]: Wrong name - this is what travis errored on.

File: packages/chart/src/Pie.ts [hunk 3]
```diff
@@ -102,7 +102,9 @@ export class Pie extends SVGWidget {
             .padRadius(outerRadius)
             .outerRadius(outerRadius)
             ;
-        const arc = this._slices.selectAll(".arc").data(this.d3Pie(this.data()), d => d.data[0]);
+        const pie_generator = this.getPieGenerator();
+        const pie_data = pie_generator(this.data());
+        const arc = this._slices.selectAll(".arc").data(pie_data, d => d.data[0]);
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: With changes made in other comment, these lines will not need to change.

File: packages/chart/src/HalfDonut.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        this._element.attr("transform", "translate(" + x + "," + y + ")scale(" + this._widgetScale + ")");
+    }
+    getPieGenerator() {
+        return this.d3Pie
+            .startAngle(-Math.PI / 2)
+            .endAngle(Math.PI / 2)
+            ;
+    }
+}
+HalfDonut.prototype._class += " common_HalfDonut";
```
Comment (GordonSmith) [inline packages/chart/src/HalfDonut.ts]: wrong class name (travis error).

File: packages/chart/src/HalfDonut.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            .startAngle(-Math.PI / 2)
+            .endAngle(Math.PI / 2)
+            .padAngle(0.0025)
+            .sort(function (b, a) {
+                return a[1] < b[1] ? -1 : a[1] > b[1] ? 1 : 0;
+            })
+            .value(function (d) {
+                return d[1];
+            })
+            ;
```
Comment (GordonSmith) [inline packages/chart/src/HalfDonut.ts]: If you called: `super.updateD3Pie()` at the start of this method, you would not need to re-add the `padAngle`, `sort` and `value` calls?

File: packages/chart/src/QuarterDonut.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            .endAngle(this.orientation() === "left" ? Math.PI / 2 : 0)
+            .padAngle(0.0025)
+            .sort(function (b, a) {
+                return a[1] < b[1] ? -1 : a[1] > b[1] ? 1 : 0;
+            })
+            .value(function (d) {
+                return d[1];
+            })
+            ;
+    }
```
Comment (GordonSmith) [inline packages/chart/src/QuarterDonut.ts]: If you called: `super.updateD3Pie()` at the start of this method, you would not need to re-add the `padAngle`, `sort` and `value` calls?

**Pull request merged**